File: old_baser/javarecipe/src/main/java/work/shion/javarecipe/pages/websample/presenters/MainPresenter.java
package work.shion.javarecipe.pages.websample.presenters;

import android.text.TextUtils;
import android.webkit.WebChromeClient;
import android.webkit.WebViewClient;

import com.annimon.stream.Optional;

import java.lang.ref.WeakReference;

import work.shion.baser.android.WebChromeClientBuilder;
import work.shion.baser.android.WebViewBuilder;
import work.shion.baser.android.WebViewClientBuilder;
import work.shion.javarecipe.pages.websample.contracts.MainPresenterContract;
import work.shion.javarecipe.pages.websample.contracts.MainViewContract;


/**
 * メインコンテンツの挙動実装
 */
public class MainPresenter implements MainPresenterContract {

    private final WeakReference<MainViewContract> viewer;


    public MainPresenter(
            WeakReference<MainViewContract> viewer
    ) {
        this.viewer = viewer;
    }


    /**
     * 最初のWeb ページ呼び出し
     */
    @Override
    public void callFirstWebPage() {
        WebChromeClient chromeClient = new WebChromeClientBuilder()
                .setOnJsAlert((view, url, message, result) -> {
                    if (viewer.get() != null && !TextUtils.isEmpty(message)) {
                        viewer.get().showAlert(message);
                        result.confirm();
                        return true;
                    } else {
                        return false;
                    }
                })
                .build();

        WebViewClient client = new WebViewClientBuilder()
                .setShouldOverrideUrlLoading((view, request) -> {
                    String scheme = request.getUrl().getScheme();
                    if (scheme != null && scheme.equalsIgnoreCase("file")) {
                        // WebView 内で表示させる
                        return false;
                    } else {
                        // file プロトコルでないものはアプリ外ブラウザで表示する
                        Optional.ofNullable(
                                this.viewer.get()
                        ).ifPresent(viewer ->
                                viewer.goBrowser(request.getUrl())
                        );
                        return true;
                    }
                })
                .build();

        Optional.ofNullable(viewer.get())
                .ifPresent(viewer -> {
                    viewer.setupWebView(new WebViewBuilder()
                            .setJavaScriptEnabled(true)
                            .setUserAgentString("偽装されたUserAgent")
                            .setWebChromeClient(chromeClient)
                            .setWebViewClient(client)
                    );
                });
    }
}

File: ktrecipe/src/main/java/work/shion/ktrecipe/pages/websample/presenters/MainPresenter.kt
package work.shion.ktrecipe.pages.websample.presenters

import work.shion.baser.android.WebChromeClientBuilder
import work.shion.baser.android.WebViewBuilder
import work.shion.baser.android.WebViewClientBuilder
import work.shion.ktrecipe.pages.websample.contracts.MainPresenterContract
import work.shion.ktrecipe.pages.websample.contracts.MainViewContract
import java.lang.ref.WeakReference


/**
 * メインコンテンツの挙動実装
 */
internal class MainPresenter(
        private val viewer: WeakReference<MainViewContract>
) : MainPresenterContract {

    /**
     * 最初のWeb ページ呼び出し
     */
    override fun callFirstWebPage() {
        viewer.get()?.setupWebView(WebViewBuilder()
                .setJavaScriptEnabled(true)
                .setUserAgentString("偽装されたUserAgent")
                .setWebChromeClient(WebChromeClientBuilder()
                        .setOnJsAlert { view, url, message, result ->
                            val viewer = viewer.get() ?: return@setOnJsAlert false
                            if (!message.isNullOrEmpty()) {
                                viewer.showAlert(message)
                                result?.confirm()
                                true
                            } else {
                                false
                            }
                        }
                        .build()
                )
                .setWebViewClient(WebViewClientBuilder()
                        .setShouldOverrideUrlLoading { view, request ->
                            val viewer = viewer.get() ?: return@setShouldOverrideUrlLoading true
                            when (request?.url?.scheme?.equals("file", ignoreCase = true)) {
                                // WebView 内で表示させる
                                true -> false

                                // file プロトコルでないものはアプリ外ブラウザで表示する
                                else -> {
                                    viewer.goBrowser(request?.url!!)
                                    true
                                }
                            }
                        }
                        .build()
                )
        )
    }
}
